PIPS/pip-1.md

```diff
@@ -0,0 +1,133 @@
+
+    // `hash` is any supported hashing function of the respective blockchain
+    bytes32 key = hash(assetHash, nativeChainId, nativeLockProxy, nativeAssetHash);
+
+    require(balances[key] == 0);
+    balances[key] = totalSupply;
+
+    delegatedAssets[_msgSender()] = DelegatedAsset(nativeChainId, nativeLockProxy, nativeAssetHash);
+}
+```
+
+This mapping will be used in the `unlock` function to ensure that delegated tokens will be unlocked only if the correct corresponding token is locked.
+
+`setManagerProxy` can be restricted to be called once. After it is called the first time, the `managerProxyContract` address cannot be changed.
+
+`bindProxyHash` and `bindAssetHash` functions can be removed.
+
+The `lock` function can be modified to require the following parameters:
+- address fromAssetHash
+- uint64 toChainId
+- bytes memory targetProxyHash
+- bytes memory toAssetHash
+- bytes memory toAddress
+- uint256 amount
+
+The LockProxy contract should maintain a `balances` mapping of (bytes32 => uint256).
+This `balances` mapping should be updated within the `lock` function:
+```
+// help prevent user mistakes by ensuring delegated assets only get transferred
+// to the native chainId, LockProxy and assetHash that was originally specified
+DelegatedAsset dAsset = delegatedAssets[fromAssetHash];
+if (dAsset.nativeChainId != 0) {
+    require(
+        dAsset.nativeChainId == toChainId &&
+        dAsset.nativeLockProxy == targetProxyHash &&
+        dAsset.nativeAssetHash == toAssetHash
+    );
+}
+
+bytes32 key = hash(fromAssetHash, toChainId, targetProxyHash, toAssetHash);
+
+balances[key] += amount;
+
+address fromContractAddr = address(this);
+TxArgs memory txArgs = TxArgs({
+    fromContractAddr: fromContractAddr,
+    fromAssetHash: fromAssetHash,
+    toAssetHash: toAssetHash,
+    toAddress: toAddress,
+    amount: amount
+});
+bytes memory txData = _serializeTxArgs(txArgs);
+
+emit LockEvent(fromContractAddr, fromAssetHash, toChainId, toAssetHash, toAddress, amount);
+```
+
+The `unlock` function should be modified to require the following parameters:
+- bytes memory argsBs
+
+Where `argsBs` contains `fromContractAddr`, `fromAssetHash`, `fromChainId`, `toAssetHash`, `toAddress` and `amount`.
+
+The contract should check that the balances are sufficient, then reduce the balance and perform the unlock:
+```
+TxArgs memory args = _deserializTxArgs(argsBs);
+DelegatedAsset dAsset = delegatedAssets[args.toAssetHash];
+
+if (dAsset.nativeChainId != 0) {
+    require(
+        dAsset.nativeChainId == args.fromChainId &&
+        dAsset.nativeLockProxy == args.fromContractAddr &&
+        dAsset.nativeAssetHash == args.fromAssetHash
+    );
+}
+
+bytes32 key = hash(args.toAssetHash, args.fromChainId, args.fromContractAddr, args.fromAssetHash);
+require(balances[key] >= args.amount);
+balances[key] -= args.amount;
+
+emit UnlockEvent(args.fromChainId, args.fromContractAddr, args.fromAssetHash, args.toAssetHash, args.toAddress, args.amount);
```

Comment:
Suggest the event follows the format of UnlockEvent(address toAssetHash, address toAddress, uint256 amount, ...optional). The reason behind that is polynetwork explorer will support display of the cross chain asset transaction details (currently including toAssetHash, toAddress, amount), and the ahead three fields need to be toAssetHash, toAddress, and amount to make the event be recognized by the explorer. As for fromChainId, fromContractAddr, fromAssetHash or any other info, can be put at the optional position.

Reply:
sure, will change it